src/Domain/Services/PayForOrder.php

```diff
@@ -0,0 +1,51 @@
+<?php
+namespace Automattic\WooCommerce\Blocks\Domain\Services;
+
+/**
+ * Service class implementing authentication checks for pay-for-order checkout block.
+ */
+class PayForOrder {
+
+	/**
+	 * Constructor.
+	 */
+	public function __construct() {
+		$this->init();
+	}
+
+	/**
+	 * Hook into WP.
+	 */
+	public function init() {
+		add_filter( 'the_content', array( $this, 'check_authentication' ) );
+	}
+
+	/**
+	 * Check authentication for pay for order
+	 *
+	 * @param string $content Content.
+	 */
+	public function check_authentication( $content ) {
+		global $wp;
+
+		$order_id = ! empty( $wp->query_vars['order-pay'] ) ? $wp->query_vars['order-pay'] : null;
+
+		if ( isset( $_GET['pay_for_order'], $_GET['key'] ) && $order_id ) { // phpcs:ignore WordPress.Security.NonceVerification
+			$order = wc_get_order( $order_id );
+
+			// Logged out customer does not have permission to pay for this order.
+			if ( ! current_user_can( 'pay_for_order', $order_id ) && ! is_user_logged_in() ) {
```

Comment:
It looks like the `current_user_can` logic already handles user ID validation, and allows guests to pay for guest orders.

Reply:
This check could also move above the `wc_get_order` to prevent loading an order if its not needed.